.github/workflows/podvm_mkosi.yaml

```diff
@@ -19,81 +46,140 @@ defaults:
     permissions:
       contents: read
       packages: write
     steps:
       - name: Checkout
         uses: actions/checkout@v4
+        with:
+          fetch-depth: 0
+          ref: "${{ inputs.git_ref }}"
+
+      - name: Rebase the code
+        if: github.event_name == 'pull_request_target'
+        working-directory: ./
+        run: |
+          ./hack/ci-helper.sh rebase-atop-of-the-latest-target-branch
 
       - name: Set up Docker Buildx
         uses: docker/setup-buildx-action@v3
 
-      - name: Log in to GitHub Container Registry
+      - name: Login to quay Container Registry
+        if: ${{ startsWith(inputs.registry, 'quay.io') }}
+        uses: docker/login-action@v3
+        with:
+          registry: quay.io
+          username: ${{ secrets.QUAY_USERNAME }}
+          password: ${{ secrets.QUAY_PASSWORD }}
+
+      - name: Login to the ghcr Container registry
+        if: ${{ startsWith(inputs.registry, 'ghcr.io') }}
         uses: docker/login-action@v3
         with:
           registry: ghcr.io
           username: ${{ github.actor }}
           password: ${{ secrets.GITHUB_TOKEN }}
 
-      - name: Build builder
-        uses: docker/build-push-action@v6
+      - name: Install build dependencies
+        run: |
+          sudo apt-get update -y
+          sudo apt-get install -y bubblewrap alien dnf qemu-utils uidmap
+          sudo snap install yq
+
+      - name: Read properties from versions.yaml
+        if: ${{ inputs.arch == 's390x' }}
+        run: |
+          go_version="$(yq '.tools.golang' versions.yaml)"
+          echo "GO_VERSION=${go_version}" >> "$GITHUB_ENV"
+
+      - name: Setup Golang version ${{ env.GO_VERSION }}
+        if: ${{ inputs.arch == 's390x' }}
+        uses: actions/setup-go@v5
         with:
-          tags: ghcr.io/${{ github.repository }}/podvm/builder-fedora:${{ github.sha }}
-          context: src
-          file: src/cloud-api-adaptor/podvm/Dockerfile.podvm_builder.fedora
-          push: true
+          go-version: ${{ env.GO_VERSION }}
 
-      # Build binaries need git commit id as part of image name
-      - name: Prepare .git folder
-        working-directory: ./
+      - name: Build builder
+        id: build_builder
+        working-directory: src/cloud-api-adaptor/podvm-mkosi
         run: |
-          cp -rf .git src/.git
+          # If the builder image matching our git sha exists, then skip this build
+          builder_image=${{ inputs.registry }}/podvm-builder-fedora-${{ inputs.arch }}
+          tag=${{ github.sha }}
+
+          if ! docker manifest inspect ${builder_image}:${tag}; then
+            PODVM_BUILDER_IMAGE="${builder_image}:${tag}" make fedora-binaries-builder
+          fi
+          echo "image=${builder_image}:${tag}" | tee -a "$GITHUB_OUTPUT"
+
+          # If the input has a different image-tag then also push it with that tag
+          if [ -n "${{ inputs.image_tag }}" ] && [ "${{ inputs.image_tag }}" != ${{ github.sha }} ];then
```

Comment:
Next commit enables `make fedora-binaries-builder` to also push the image, based on value of `PUSH` envvar. I glanced over all commits, it seems that `PUSH` is always `true`. In future it might be `false` on depending on, for example, a workflow input. So I think this `if` should also check the value of `PUSH`.

On a related matter, as is this workflow it will always push the image. Even on pull requests, transient (disposable) images will be pushed. Today we already push images for pull requests, but these are easily traceable by tag (`pr-number`)  which allow us to run clean up routines like in https://github.com/confidential-containers/cloud-api-adaptor/pull/1443 . When we push transient images tagged by sha it's harder to do removal. I'm not saying we should care, in fact I'm okay with pushing all transient images; I just wanted to raise that for general awareness.

Reply:
Yeah, I deliberately chose to cache the binaries images as they can take quite a long time to build (though with the caching of components maybe this isn't a problem anymore) and In the previous flow there was some logic to pass through a prior binaries build to speed this up, which I didn't think was helpful.

I'm not completely sure if there is a problem with having more and more tags in ghcr (the kata-agent cache for example has ~13.5 tags), but if there is maybe we can turn PUSH off for PR builds if you'd prefer, or prefix the tag with the pr number?